[labs/analyzer] Declared union types should be preserved

### Which package(s) are affected?

Analyzer (@lit-labs/analyzer)

### Description

This is crucial especially for generated Angular wrappers.

### Reproduction

Consider a property like

```ts
@property({ type: String, reflect: true })
public placement: 'top' | 'bottom' = 'top';
```

Currently, the analyzer would widen its type to `string`. Then, the Angular wrapper generator would emit

```ts
@Input()
set placement(v: string) {
  this._ngZone.runOutsideAngular(() => (this._el.placement = v));
}
```

which fails type checking:

```
error TS2322: Type 'string' is not assignable to type '"top" | "bottom"'.
```


### Workaround

Fixed by this three-line change, which we've successfully run in production for a year:
```diff
diff --git a/packages/labs/analyzer/src/lib/types.ts b/packages/labs/analyzer/src/lib/types.ts
index 803eb141..f6b7442f 100644
--- a/packages/labs/analyzer/src/lib/types.ts
+++ b/packages/labs/analyzer/src/lib/types.ts
@@ -84,8 +84,11 @@ export const getTypeForType = (
 ): Type => {
   const {typescript} = analyzer;
   const checker = analyzer.program.getTypeChecker();
-  // Ensure we treat inferred `foo = 'hi'` as 'string' not '"hi"'
-  type = checker.getBaseTypeOfLiteralType(type);
+  // Ensure we treat inferred `foo = 'hi'` as 'string' not '"hi"',
+  // but retain explicitly defined unions
+  if (!type.isUnion()) {
+    type = checker.getBaseTypeOfLiteralType(type);
+  }
   const text = checker.typeToString(type);
   const typeNode = checker.typeToTypeNode(
     type,
```

### Is this a regression?

No or unsure. This never worked, or I haven't tried before.

### Affected versions

0.13.2

### Browser/OS/Node environment

OS: Windows 11
Node version: v25.2.1
npm version: 11.6.2

--- a/.changeset/honest-planes-send.md
+++ b/.changeset/honest-planes-send.md
@@ -1,0 +1,5 @@
+---
+'@lit-labs/analyzer': minor
+---
+
+Declared union types are now preserved instead of being widened to their base type. This fixes unexpected type errors in the property accessors created by the Angular wrapper generator.

--- a/packages/labs/analyzer/src/lib/types.ts
+++ b/packages/labs/analyzer/src/lib/types.ts
@@ -84,8 +84,11 @@ export const getTypeForType = (
 ): Type => {
   const {typescript} = analyzer;
   const checker = analyzer.program.getTypeChecker();
-  // Ensure we treat inferred `foo = 'hi'` as 'string' not '"hi"'
-  type = checker.getBaseTypeOfLiteralType(type);
+  // Ensure we treat inferred `foo = 'hi'` as 'string' not '"hi"',
+  // but retain explicitly defined unions
+  if (!type.isUnion()) {
+    type = checker.getBaseTypeOfLiteralType(type);
+  }
   const text = checker.typeToString(type);
   const typeNode = checker.typeToTypeNode(
     type,

--- a/packages/labs/analyzer/src/test/server/lit/properties_test.ts
+++ b/packages/labs/analyzer/src/test/server/lit/properties_test.ts
@@ -175,6 +175,22 @@ for (const lang of languages) {
       assert.equal(importedClass.module, 'external.js');
     });
 
+    test('property with string literal union', () => {
+      const property = element.reactiveProperties.get('stringLiteralUnion')!;
+      assert.equal(property.type?.text, '"hi" | "hello"');
+      assert.equal(property.type?.references.length, 0);
+      assert.equal(property.reflect, false);
+    });
+
+    test('reflected property with string literal union', () => {
+      const property = element.reactiveProperties.get(
+        'reflectedStringLiteralUnion'
+      )!;
+      assert.equal(property.type?.text, '"hi" | "hello"');
+      assert.equal(property.type?.references.length, 0);
+      assert.equal(property.reflect, true);
+    });
+
     test('reflect: true', () => {
       const property = element.reactiveProperties.get('reflectTrue')!;
       assert.equal(property.reflect, true);

--- a/packages/labs/analyzer/src/test/server/types_test.ts
+++ b/packages/labs/analyzer/src/test/server/types_test.ts
@@ -116,6 +116,18 @@ for (const lang of languages) {
       assert.equal(type.references[1].isGlobal, true);
     });
 
+    test('testStringLiteralUnion', () => {
+      const type = typeForVariable(module, 'testStringLiteralUnion');
+      assert.equal(type.text, '"hi" | "hello"');
+      assert.equal(type.references.length, 0);
+    });
+
+    test('inferredStringLiteralUnion', () => {
+      const type = typeForVariable(module, 'inferredStringLiteralUnion');
+      assert.equal(type.text, '"hi" | "hello"');
+      assert.equal(type.references.length, 0);
+    });
+
     test('inferredLocalClass', () => {
       const type = typeForVariable(module, 'inferredLocalClass');
       assert.equal(type.text, 'LocalClass');

--- a/packages/labs/analyzer/test-files/js/properties/element-a.js
+++ b/packages/labs/analyzer/test-files/js/properties/element-a.js
@@ -34,6 +34,8 @@ export class ElementA extends LitElement {
     importedClass: {},
     globalClass: {},
     union: {},
+    stringLiteralUnion: {},
+    reflectedStringLiteralUnion: {type: String, reflect: true},
     staticProp: {attribute: 'static-prop', type: Number},
   };
 
@@ -75,6 +77,8 @@ export class ElementA extends LitElement {
     this.localClass = new LocalClass();
     this.importedClass = new ImportedClass();
     this.globalClass = document.createElement('foo');
+    this.stringLiteralUnion = /** @type {'hi' | 'hello'} */ ('hi');
+    this.reflectedStringLiteralUnion = /** @type {'hi' | 'hello'} */ ('hi');
     this.staticProp = 42;
   }
 

--- a/packages/labs/analyzer/test-files/js/types/module.js
+++ b/packages/labs/analyzer/test-files/js/types/module.js
@@ -40,6 +40,10 @@ export const testStringImportedClassUnion = 'hi';
 /** @type {string | ImportedClass | HTMLElement} */
 export const testStringImportedGlobalClassUnion = 'hi';
 
+/** @type {'hi' | 'hello'} */
+export const testStringLiteralUnion = 'hi';
+export const inferredStringLiteralUnion = /** @type {'hi' | 'hello'} */ ('hi');
+
 export const inferredLocalClass = new LocalClass();
 export const inferredImportedClass = new ImportedClass();
 export const inferredExternalClass = new LitElement();

--- a/packages/labs/analyzer/test-files/ts/properties/src/element-a.ts
+++ b/packages/labs/analyzer/test-files/ts/properties/src/element-a.ts
@@ -99,6 +99,12 @@ export class ElementA extends LitElement {
   @property()
   union: LocalClass | HTMLElement | ImportedClass;
 
+  @property()
+  stringLiteralUnion: 'hi' | 'hello';
+
+  @property({type: String, reflect: true})
+  reflectedStringLiteralUnion: 'hi' | 'hello';
+
   @state()
   private _stateField: number;
 

--- a/packages/labs/analyzer/test-files/ts/types/src/module.ts
+++ b/packages/labs/analyzer/test-files/ts/types/src/module.ts
@@ -31,6 +31,9 @@ export const testStringImportedGlobalClassUnion:
   | ImportedClass
   | HTMLElement = 'hi';
 
+export const testStringLiteralUnion: 'hi' | 'hello' = 'hi';
+export const inferredStringLiteralUnion = 'hi' as 'hi' | 'hello';
+
 export const inferredLocalClass = new LocalClass();
 export const inferredImportedClass = new ImportedClass();
 export const inferredExternalClass = new LitElement();

